src/sdam/description/topology/mod.rs

```diff
@@ -532,17 +566,24 @@ impl TopologyDescription {
     }
 
     /// Create a new ServerDescription for each address and add it to the topology.
-    fn add_new_servers<'a>(&'a mut self, servers: impl Iterator<Item = &'a String>) -> Result<()> {
-        for server in servers {
-            let server = StreamAddress::parse(&server)?;
+    fn add_new_servers<'a>(&mut self, servers: impl Iterator<Item = &'a String>) -> Result<()> {
+        let servers: Result<Vec<_>> = servers.map(|server| StreamAddress::parse(server)).collect();
+
+        self.add_new_servers_from_addresses(servers?.iter());
+        Ok(())
+    }
 
+    /// Create a new ServerDescription for each address and add it to the topology.
+    fn add_new_servers_from_addresses<'a>(
```

Comment:
Since I needed a subset of the functionality of `add_new_servers` for `sync_hosts`, I split it out into a separate method